### ISSUE TITLE
[MSAL for Angular] Support for Angular 6/7/8

### ISSUE THREAD

**Issue by anth-git** [AzureAD/microsoft-authentication-library-for-js#421]
## I'm submitting a...
<pre><code>
[x] Feature request
</code></pre>
## Current behavior
No support for Angular 6
## Expected behavior
Support for Angular 6 ;)
</code></pre>

In #398 you stated that
> We don't have support for angular 6 yet but that's in our backlog and we will work on it in future.

I would like to ask if you have any rough estimate when can we expect it?
Is it a matter of weeks or months?

**Comment by BenBullock1992** [comment]
@anth-git - While we wait for Angular 6 support, if it helps, I managed to get this working by simply installed the rxjs-compat npm module. This resolved all the RxJS peer issues I was experiencing.

**Comment by anth-git** [comment]
@BenBullock1992 Thanks a lot, it indeed works now, although I still have peer dependency warnings:

`warning " > @azure/msal-angular@0.1.1" has incorrect peer dependency "@angular/common@^4.3.0".
warning " > @azure/msal-angular@0.1.1" has incorrect peer dependency "@angular/core@^4.3.0".
warning " > @azure/msal-angular@0.1.1" has incorrect peer dependency "rxjs@^5.0.1".`

**Comment by BenBullock1992** [comment]
@anth-git - Awesome, good to hear! And yeah, I get the warnings too. I suspect when Microsoft support MSAL for Angular 6 these warnings will go away.

**Comment by nehaagrawal** [comment]
@anth-git @BenBullock1992  We are planning to give support for angular 6 in future. As @BenBullock1992  suggested you should be able to make it work by using rxjs-compat for now. I will update this ticket once we release support for angular 6.

**Comment by anth-git** [comment]
@rsuk I have the same error, but strangely enough only after first API call. Subsequent calls are handled correctly

**Comment by RobARichardson** [comment]
@rsuk @anth-git I'm having the exact same issue.

**Comment by anth-git** [comment]
For now I resolved it by implementing my own interceptor based on MsalInterceptor

**Comment by BenBullock1992** [comment]
@anth-git - I'm not experiencing this at runtime but am getting the error when running tests. Can you post your interceptor code for reference please?

**Comment by anth-git** [comment]
It's just copy-pasted MsalInterceptor, with replaced Observable.fromPromise to from

`import { HttpErrorResponse, HttpEvent, HttpHandler, HttpInterceptor, HttpRequest } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { BroadcastService, MsalService } from '@azure/msal-angular';
import { from } from 'rxjs';
import { Observable } from 'rxjs/Observable';

// TODO: Remove when MSAL library supports Angular 6, and use MsalInterceptor instead

@Injectable()
export class MsalInterceptorTmp implements HttpInterceptor {

    constructor(private auth: MsalService, private broadcastService: BroadcastService) { }

    intercept(req: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
        const scopes = this.auth.getScopesForEndpoint(req.url);
        this.auth.verbose('Url: ' + req.url + ' maps to scopes: ' + scopes);
        if (scopes === null) {
            return next.handle(req);
        }
        const tokenStored = this.auth.getCachedTokenInternal(scopes);
        if (tokenStored && tokenStored.token) {
            req = req.clone({
                setHeaders: {
                    Authorization: `Bearer ${tokenStored.token}`,
                }
            });
            return next.handle(req).do(event => { }, err => {
                if (err instanceof HttpErrorResponse && err.status === 401) {
                    const scopesl = this.auth.getScopesForEndpoint(req.url);
                    const tokenStoredl = this.auth.getCachedTokenInternal(scopesl);
                    if (tokenStoredl && tokenStoredl.token) {
                        this.auth.clearCacheForScope(tokenStoredl.token);
                    }
                    this.broadcastService.broadcast('msal:notAuthorized', { err, scopesl });
                }
            });
        } else {
            return from(this.auth.acquireTokenSilent(scopes).then(token => {
                const JWT = `Bearer ${token}`;
                return req.clone({
                    setHeaders: {
                        Authorization: JWT,
                    },
                });
            })).mergeMap(reql => next.handle(reql).do(event => { }, err => {
                if (err instanceof HttpErrorResponse && err.status === 401) {
                    const scopesl = this.auth.getScopesForEndpoint(reql.url);
                    const tokenStoredl = this.auth.getCachedTokenInternal(scopesl);
                    if (tokenStoredl && tokenStoredl.token) {
                        this.auth.clearCacheForScope(tokenStoredl.token);
                    }
                    this.broadcastService.broadcast('msal:notAuthorized', { err, scopesl });
                }
            })); // calling next.handle means we are passing control to next interceptor in chain
        }
    }
}
`

**Comment by RobARichardson** [comment]
Thanks, @anth-git. The interceptor works for me till this gets fixed.

**Comment by rajarameshvarma** [comment]
angular 7 is out. Is this still work in progress?

**Comment by 217Wilder-User1** [comment]
> angular 7 is out. Is this still work in progress?

@rajarameshvarma I'm using "@angular/core": "~7.0.0" and it works in Edge and Chrome.

**Comment by pandres95** [comment]
> angular 7 is out. Is this still work in progress?

Is this still work in progress?

**Comment by Dedme** [comment]
> > angular 7 is out. Is this still work in progress?
> 
> @rajarameshvarma I'm using "@angular/core": "~7.0.0" and it works in Edge and Chrome.

i also have this built and working in angular 7.1.0. but the above fix was required.
i am using patch-package to change the dist files for the time being.

**Comment by pandres95** [comment]
@Dedme, is there a PR patching that file up?

**Comment by Dedme** [comment]
@pandres95, i dont believe so.

**Comment by Dedme** [comment]
@pandres95 its unnecessarily large but attached is my package for patching the current npm file:
https://gist.github.com/Dedme/919d85c0390c7a82dfdacfa4d29881f9

this is applied using https://www.npmjs.com/package/patch-package.
this patch has the compiled codebase to fix IE compatibility and also the above fix resolving the observable.

**Comment by DeanB2015** [comment]
@Dedme Thanks for the .patch file.  Just to confirm.  The .patch is a es2015 dist of `@azure\msal-angular`?

Note: I've posted about the IE11 issues here:
https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/414  Unfortunately, the es2015 dist files are causing problems with Edge.  Do you get errors in Edge now?

**Comment by Dedme** [comment]
@DeanB2015 with my current patch file, i am having no issues with edge.

I have noted while it compiles, loads and runs in IE11, it is not authenticating via the popup,
it is throwing some sort of error.
apologies, out target browser has been Chrome.

**Comment by Raysapps** [comment]
I got this error: "Token renewal operation failed due to timeout|Token Renewal Failed "
 in my angular 7 asp.netcore 2.2 project. Any suggestions to fix will be a great help.

**Comment by ssougnez** [comment]
When is this going to be released? It's a pretty annoying bug...

**Comment by ef32** [comment]
When is this going to be released, just an estimated time please? Angular is now on 7 and plans are in place to release 8 soon. Some feed back will be nice.

**Comment by mirkodelgado** [comment]
When is Angular 7 support going to be released? Could you please provide an estimated time frame?

**Comment by sameerag** [comment]
@mirkodelgado This upcoming summer, [roadmap](https://github.com/AzureAD/microsoft-authentication-library-for-js/wiki#roadmap) will be updated with details as we start this activity.

**Comment by souvikbasu** [comment]
Any work around for Angular 7 if this PR is not yet released? I need to authenticate to Azure AD on my Angular 7 app

**Comment by anth-git** [comment]
@souvikbasu it still works with Angular 7 and 8, but with Ivy renderer disabled.

**Comment by souvikbasu** [comment]
Just an update from my side, it works for me if rxjs-compat is installed as somebody already mentioned above. Tested on Angular 7.2.3

**Comment by kissferenc** [comment]
@souvikbasu thx, the install of rxjs-compat is works for me.

**Comment by sameerag** [comment]
#786 to track this

**Comment by dotnetcanuck** [comment]
@anth-git @souvikbasu @kissferenc The upgrade guide for Angular 8 says to remove `rxjs-compat` though...?

**Comment by kissferenc** [comment]
@dotnetcanuck I removed the rxjs-compat library and the compiler can't compile my project.
Maybe I have to remove/upgrade some deprecated methodes.

These are the error codes:

`ERROR in node_modules/rxjs/Subscription.d.ts(1,15): error TS2307: Cannot find module 'rxjs-compat/Subscription'.
node_modules/rxjs/BehaviorSubject.d.ts(1,15): error TS2307: Cannot find module 'rxjs-compat/BehaviorSubject'.
node_modules/rxjs/Observable.d.ts(1,15): error TS2307: Cannot find module 'rxjs-compat/Observable'.
node_modules/@azure/msal-angular/dist/broadcast.service.d.ts(1,10): error TS2305: Module '"../../../rxjs/Subscription"' has no exported member 'Subscription'.
node_modules/@azure/msal-angular/dist/broadcast.service.d.ts(4,10): error TS2305: Module '"../../../rxjs/BehaviorSubject"' has no exported member 'BehaviorSubject'.
node_modules/@azure/msal-angular/dist/broadcast.service.d.ts(5,10): error TS2305: Module '"../../../rxjs/Observable"' has no exported member 'Observable'.
node_modules/@azure/msal-angular/dist/msal.interceptor.d.ts(2,10): error TS2305: Module '"../../../rxjs/Observable"' has no exported member 'Observable'.
`

**Comment by baltie** [comment]
You shouldn't remove the rxjs-compat libary until MSAL for Angular version 1.0 is ready. It is needed by the current version 0.1.2.

**Comment by dotnetcanuck** [comment]
@kissferenc @baltie I don't really understand how. Our project isn't using `rxjs-compat` at all, and MSAL has been working fine so far. That's why I am so confused.

The project has been on Angular 7.3 until now, and I just upgraded it to Angular 8. I removed the MSAL package to get past the errors thrown by `ng update` (although later I found the `--force` flag), and then reinstalled it after the upgrade was done.

The project is still working fine, running on Angular 8, RxJS 6.5.2, and no references to `rxjs-compat` whatsoever.  Everything I am reading says that `rxjs-compat` should be removed once you're on RxJS 6, so why is it that my project is working fine, yet, multiple people in this thread are seeing/saying the opposite?

**Comment by baltie** [comment]
This is the MSAL package you're using? https://www.npmjs.com/package/@azure/msal-angular

It depends on RxJS 5. So if it is working without you explicitly installing rxjs-compat, something else you're using is installing it for you. Have you tried `npm ls rxjs-compat` in your project directory? You should be able to see if it has been added by some other package. You can also check node_modules folder to see if the rxjs-compat folder is there with files in it. What I'm completely sure is that version 0.1.2 of the @azure/msal-angular package does not work with RxJS 6.5.2 without rxjs-compat.

For your last statement, you should remove rxjs-compat when your code has been upgraded to use RxJS 6. Since MSAL for Angular hasn't been upgraded to RxJS 6 and doesn't have rxjs-compat in its dependency list, you have to add it yourself.

**Comment by dotnetcanuck** [comment]
@baltie That is indeed the package that I've installed, yes.

Everything that you are saying makes logical sense to me, because a package that depends on RxJS 5 will not run against RxJS 6 without rxjs-compat. But that's the total opposite of what I'm seeing. Even our CI builds in DevOps haven't thrown any errors, and when I deploy the site to our Azure dev environment, everything works! (and yes, I used incognito mode, so no caching)

Here's the output of `npm ls` for both rxjs-compat and msal-angular, exact paths/names redacted, of course. 😛 

```
$ npm ls rxjs-compat
my-app@0.0.0 C:\Git\repo\
`-- (empty)
```

```
$ npm ls @azure/msal-angular
my-app@0.0.0 C:\Git\repo\
`-- @azure/msal-angular@0.1.2
```

What's weirder still is that when I run `npm ls rxjs`, the peer dependency error shows up, but the app compiles and runs.

```
$ npm ls rxjs
my-app@0.0.0 C:\Git\repo
+-- @angular-devkit/build-angular@0.800.4
| +-- @angular-devkit/architect@0.800.4
| | `-- rxjs@6.4.0
| +-- @angular-devkit/build-webpack@0.800.4
| | `-- rxjs@6.4.0
| +-- @angular-devkit/core@8.0.4
| | `-- rxjs@6.4.0
| +-- @ngtools/webpack@8.0.4
| | `-- rxjs@6.4.0
| `-- rxjs@6.4.0
+-- @angular/cli@8.0.4
| +-- @angular-devkit/schematics@8.0.4
| | `-- rxjs@6.4.0
| +-- @schematics/update@0.800.4
| | `-- rxjs@6.4.0
| `-- inquirer@6.3.1
|   `-- UNMET PEER DEPENDENCY rxjs@6.5.2  deduped
+-- my-other-custom-library@0.1.7-beta1
| `-- rxjs@6.3.3
`-- UNMET PEER DEPENDENCY rxjs@6.5.2

npm ERR! peer dep missing: rxjs@^5.0.1, required by @azure/msal-angular@0.1.2
npm ERR! peer dep missing: rxjs@^5.0.1, required by @azure/msal-angular@0.1.2
```

We're using the MSAL package to retrieve authorization tokens from Azure B2C. So my conclusion is that either we're using functions of the MSAL library that don't seem to care about RxJS at all, or that there's some other kind of voodoo magic going on.

Either way, this project doesn't go live for a little while still, so hopefully version 1.0 will be ready before it becomes a production issue?

¯\\\_(ツ)_/¯

**Comment by dotnetcanuck** [comment]
Update... So I don't know what changed, but our app suddenly stopped working with B2C, and would redirect back to the login screen after authentication was completed.

It took us a lot of time to diagnose it, and we eventually circled back to this very issue. Adding `rxjs-compat` fixed the problem, for the moment at least. (Cue me eating my words, above)

That said... given that MSAL.js 1.0.2 has been released, is there an updated timeline on when we can expect an update to `msal-angular`? Or perhaps have one of the several open pull requests (e.g. [PR 543](https://github.com/AzureAD/microsoft-authentication-library-for-js/pull/543)) merged, even as a temporary Preview? @sameerag (I would have asked in #786, but it's locked to collaborators only)

**Comment by sameerag** [comment]
@dotnetcanuck Let me get back on the temp preview, we are currently in planning for this activity and can do small incremental fixes if they help unblock some folks. cc @DarylThayil. 

#786 is locked for traction, easier to have all issues bunched up. We will open a feedback PR for Angular once we have our planning locked.

**Comment by Yossri21** [comment]
I am using Angular 8 &  simply installing  ' npm i rxjs-compat ' that's handle all the problems .

**Comment by GirishGowda37** [comment]
does MSAL supports in angular  7 version.? 
I am currently using ADAL for authenticating  with AD.
Please let me know

**Comment by DarylThayil** [comment]
we do not have stated support for angular 7 yet, we hope to have that by the end of the year.
Some have found ways to make this work for angular 7+ 
cc @jasonnutter

**Comment by davidgg** [comment]
I'm using it with Angular 8. Just install `rxjs-compat` and configure the module with your settings.

**Comment by jasonnutter** [comment]
As noted, we are working now to provide out of the box support for the newer versions of Angular (planning to have this available by the end of the year). In the mean time, we have heard that you can use `@azure/msal-angular` with the newer versions of Angular with some changes in your application (e.g. as @davidgg notes, using `rxj-compat`).

**Comment by ShinRai1090** [comment]
> As noted, we are working now to provide out of the box support for the newer versions of Angular (planning to have this available by the end of the year). In the mean time, we have heard that you can use `@azure/msal-angular` with the newer versions of Angular with some changes in your application (e.g. as @davidgg notes, using `rxj-compat`).

Hope this comes out sooner, angular 9 is already almost here and even though there is support for latest versions using _rxjs-compat_, encountering below issues while performing a simple "_ng update @angular/cli_"

`
Fetching dependency metadata from registry...
                  Package "@azure/msal-angular" has an incompatible peer dependency to "@angular/common" (requires "^4.3.0" (extended), would install "8.2.14").
                  Package "@azure/msal-angular" has an incompatible peer dependency to "@angular/core" (requires "^4.3.0" (extended), would install "8.2.14").
                  Package "@azure/msal-angular" has an incompatible peer dependency to "rxjs" (requires "^5.0.1", would install "6.5.3").
Incompatible peer dependencies found.
Peer dependency warnings when installing dependencies means that those dependencies might not work correctly together.        
You can use the '--force' option to ignore incompatible peer dependencies and instead address these warnings later.
`

**Comment by Taats** [comment]
I'm building an app and I need to connect with B2C, is there any ETA on this new version because we are in the 20's now and a new version is not yet to be found. 

An estimate would be greatly appreciated!

**Comment by jasonnutter** [comment]
@Taats We're a little behind but we'll have something soon (this month), apologies for the delay!

**Comment by Taats** [comment]
@jasonnutter That is good news, thanks for the update! :)

**Comment by ShinRai1090** [comment]
@jasonnutter Can we expect the update this month?

**Comment by OlivierTD** [comment]
@jasonnutter I saw a merge of a beta version that happened today. Any ETA on when that version will be released? We're just starting a new project. It would be great if we could just use this new version right off the bat. Thanks!

**Comment by jasonnutter** [comment]
@Taats @ShinRai1090 @OlivierTD Please follow the upgrade here to try out the new beta for MSAL Angular: https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev-angular-1.0-protectedresources/lib/msal-angular/docs/0.x-1.x-upgrade-guide.md

This version will support Angular 6+ and rxjs 6.

**Comment by ShinRai1090** [comment]
> @Taats @ShinRai1090 @OlivierTD Please follow the upgrade here to try out the new beta for MSAL Angular: https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev-angular-1.0-protectedresources/lib/msal-angular/docs/0.x-1.x-upgrade-guide.md
> 
> This version will support Angular 6+ and rxjs 6. Let me know if you have any feedback, thanks! Apologies again for the delay.

@madhugore1 @mamta1992

**Comment by komi99** [comment]
getScopesForEndpoint() this method is always returning null for me  instead of returning scope , how it is reading scope from endpoint

**Comment by jasonnutter** [comment]
@komi99 Please open a new issue if you have a usage question, thanks!
